homeassistant/helpers/storage.py

```diff
@@ -88,6 +95,147 @@ def load_old_config():
     return config
 
 
+def get_internal_store_manager(
+    hass: HomeAssistant, config_dir: str | None = None
+) -> _StoreManager:
+    """Get the store manager.
+
+    This function is not part of the API and should only be
+    used in the Home Assistant core internals. It is not
+    guaranteed to be stable.
+    """
+    if STORAGE_MANAGER not in hass.data:
+        manager = _StoreManager(hass, config_dir or hass.config.config_dir)
+        hass.data[STORAGE_MANAGER] = manager
+    return hass.data[STORAGE_MANAGER]
+
+
+class _StoreManager:
+    """Class to help storing data.
+
+    The store manager is used to cache and manage storage files.
+    """
+
+    def __init__(self, hass: HomeAssistant, config_dir: str) -> None:
+        """Initialize storage manager class."""
+        self._hass = hass
+        self._invalidated: set[str] = set()
+        self._files: set[str] | None = None
+        self._data_preload: dict[str, json_util.JsonValueType] = {}
+        self._storage_path: Path = Path(config_dir).joinpath(STORAGE_DIR)
+        self._cancel_cleanup: asyncio.TimerHandle | None = None
+
+    async def async_initialize(self) -> None:
+        """Initialize the storage manager."""
+        hass = self._hass
+        await hass.async_add_executor_job(self._initialize_files)
+        hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_STARTED,
+            self._async_schedule_cleanup,
+            run_immediately=True,
+        )
+
+    @callback
+    def async_invalidate(self, key: str) -> None:
+        """Invalidate cache.
+
+        Store calls this when its going to save data
+        to ensure that the cache is not used after that.
+        """
+        if "/" not in key:
+            self._invalidated.add(key)
+            self._data_preload.pop(key, None)
+
+    @callback
+    def async_fetch(
+        self, key: str
+    ) -> tuple[bool, json_util.JsonValueType | None] | None:
+        """Fetch data from cache."""
+        #
+        # If the key is invalidated, we don't need to check the cache
+        # If async_initialize has not been called yet, we don't know
+        # if the file exists or not so its a cache miss
+        #
+        # It is very important that we check if self._files is None
+        # because we do not want to incorrectly return a cache miss
+        # because async_initialize has not been called yet as it would
+        # cause the Store to return None when it should not.
```

Comment:
I was very concerned about getting this part correct.  I added this comment to ensure we don't break it in the future.